src/main/java/spoon/support/compiler/jdt/JDTBasedSpoonCompiler.java

```diff
@@ -401,55 +380,90 @@ protected JDTBatchCompiler createBatchCompiler(InputType... types) {
 	}
 
 	protected boolean buildTemplates(JDTBuilder jdtBuilder) {
-		if (templates.getAllJavaFiles().isEmpty()) {
-			return true;
-		}
+		CompilationUnitDeclaration[] units = buildUnits(jdtBuilder, templates, getTemplateClasspath(), "template ", false);
 
-		JDTBatchCompiler batchCompiler = createBatchCompiler(InputType.FILES);
+		// here we build the model in the template factory
+		buildModel(units);
+
+		return probs.size() == 0;
+	}
+
+	private static final CompilationUnitDeclaration[] EMPTY_RESULT = new CompilationUnitDeclaration[0];
+
+	protected CompilationUnitDeclaration[] buildUnits(JDTBuilder jdtBuilder, SpoonFolder sourcesFolder, String[] classpath, String debugMessagePrefix, boolean buildOnlyOutdatedFiles) {
+		List<SpoonFile> sourceFiles = sourcesFolder.getAllJavaFiles();
+		if (sourceFiles.isEmpty()) {
+			return EMPTY_RESULT;
+		}
+		FileCompiler batchCompiler = (FileCompiler) createBatchCompiler(InputType.FILES);

```

Comment:
I do not like it too, because it forces implementations to return only that type. I have problem with createBatchCompiler method. It internaly creates a compiler which is already initialized with set of files which has to be compiled, ... but in this context we need a different set of files. I do not like it. I would prefer to split creation of compiler and initialization of set of files. But how? This method is overwritten in some tests - as example how to filter files. This stops me. 

I would prefer to create here simply the JDTBatchComplier (no FileCompiler) and then to give it the files (or directly units) which has to be compiled. But this direct aproach needs changes in the tests... @monperrus How much do you persist on FileCompiler and FactoryCompiler which comes with built-in set of files? What do you suggest?